REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [9]:
import pandas as pd

In [10]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [11]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude phone_number  
0 -122.403816  14154217044  
1 -122.421547  14157763262  
2 -122.403135  14156779494  
3 -122.419004          NaN  
4 -122.413641  14156415051

In [12]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [13]:
len(bus.index)

6315

#### Question 2b: How many unique business IDs are there?  

In [14]:
len(bus['business_id'].unique())

6315

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [15]:
bus['name'].value_counts()[:5]

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: name, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [16]:
print (bus['postal_code'].dtypes)

object


#### Question 3b

What are the unique values of postal_code?

In [17]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '0', '94188', '94621',
       '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [18]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]


In [19]:
busvalid = bus[bus['postal_code'].isin(validZip)]
len(busvalid.index)

5997

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [20]:
busnozip=busvalid[pd.isnull(busvalid['longitude'])]
len(busnozip)

2481

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [21]:
not_null_df = busvalid[['postal_code', 'longitude']].groupby('postal_code').count()
null_lon_df = busvalid[pd.isnull(busvalid['longitude'])][['postal_code', 'business_id']].groupby('postal_code').count()
postal_long = pd.merge(not_null_df, null_lon_df, how="outer", left_index=True, right_index=True)
postal_long = postal_long.rename(columns={"longitude": "not_null_lon", "business_id": "null_lon"})
postal_long

not_null_lon  null_lon
postal_code                        
94102                 262       196
94103                 306       252
94104                  74        59
94105                 126       100
94107                 204       247
94108                 142        82
94109                 237       146
94110                 334       251
94111                 167       115
94112                 131        59
94114                 122       108
94115                 162        76
94116                  58        34
94117                 130        66
94118                 143       100
94121                 106        55
94122                 147       123
94123                 116        53
94124                  76        88
94127                  43        28
94131                  34        14
94132                  69        65
94133                 282       135
94134                  45        29

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [22]:
postal_long['prop_null_lon'] = postal_long['null_lon'] / (postal_long['null_lon'] + postal_long['not_null_lon'])
postal_long.sort_values('prop_null_lon', ascending=False).head()

not_null_lon  null_lon  prop_null_lon
postal_code                                       
94107                 204       247       0.547672
94124                  76        88       0.536585
94132                  69        65       0.485075
94114                 122       108       0.469565
94122                 147       123       0.455556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [23]:
ins_bus_uni=ins['business_id'].nunique()
rows=len(ins.index)
print(ins_bus_uni,rows)
print("more than one inspection per business ")

5730 15430
more than one inspection per business 


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [24]:
ins['type'].unique()

array(['routine', 'complaint'], dtype=object)

In [25]:
ins_count=ins[['business_id','type']].groupby('type').count()
ins_count

business_id
type                  
complaint            1
routine          15429

In [26]:
ins2 = ins[ins['type']=='routine']
ins2[['business_id','type']].groupby('type').count()

business_id
type                
routine        15429

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [27]:
ins.dtypes

business_id     int64
score           int64
date            int64
type           object
dtype: object

In [31]:
ins['date']=ins['date'].astype(str)
ins['year'] = ins['date'].str[:4]
ins

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15430 rows x 5 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [32]:
ins['year'] = ins['year'].astype(int)
print(ins['year'].max())
print(ins['year'].min())
#Range is 2013 to 2016.

2016
2013


In [33]:
ins_year_count=ins[['business_id','year']].groupby('year').count()
ins_year_count
#Not the same number

business_id
year             
2013           38
2014         5629
2015         3923
2016         5840

In [34]:
year_50 = ins_year_count[ins_year_count['business_id']>50]
ins['year']=ins['year'].astype(int)
ins3 = pd.merge(ins, year_50, on='year')
ins3

business_id_x  score      date     type  year  business_id_y
0                 10     82  20160503  routine  2016           5840
1                 19     94  20160513  routine  2016           5840
2                 24     98  20161005  routine  2016           5840
3                 24     96  20160311  routine  2016           5840
4                 31     84  20160816  routine  2016           5840
...              ...    ...       ...      ...   ...            ...
15387          85766     93  20151222  routine  2015           3923
15388          85796     94  20151202  routine  2015           3923
15389          85824     74  20151208  routine  2015           3923
15390          85925     74  20151216  routine  2015           3923
15391          86344    100  20151211  routine  2015           3923

[15392 rows x 6 columns]

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [35]:
ins4=ins[(ins['year'] == 2016)]
ins4

business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5840 rows x 5 columns]

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [36]:
merge_all = pd.merge(bus, ins4, on ='business_id', how = 'left')
merge_all.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude phone_number  score      date     type    year  
0 -122.403816  14154217044   82.0  20160503  routine  2016.0  
1 -122.421547  14157763262   94.0  20160513  routine  2016.0  
2 -122.403135  14156779494   98.0  20161005  routine  2016.0  
3 -122.403135  14156779494   96.0  20160311  routine  2016.0  
4 -122.419004          NaN   84.0  20160816  routine  2016.0

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [43]:
merge_rate20_name=merge_all.sort_values('score').head(20)['name'].tolist()
for n in merge_rate20_name:
        print(n)
merge_rate_add=merge_all.sort_values('score').head(20)['address'].tolist()
for m in merge_rate_add:
        print(m)
merge_rate_rating=merge_all.sort_values('score').head(20)['score'].tolist()
for r in merge_rate_rating:
        print(r)

GOLDEN RIVER RESTAURANT
POKI TIME
CRAZY PEPPER
UNCLE CAFE
HING WANG BAKERY
GOLDEN WOK
TOP'S CAFE
L & G VIETNAMESE SANDWICH
RED A BAKERY
SAN TUNG RESTAURANT LLC
MA'S DIM SUM & CAFE
NEW GARDEN RESTAURANT, INC.
KOBANI MEDITERANEAN GRILL
JOANIE'S DINER INC.
PAK NAM THAI CUISINE
JADE GARDEN
ITHAI
VALLARTA'S TACO BAR
HONG KONG LOUNGE
WOLFES LUNCH
5827 GEARY BLVD 
2101 LOMBARD ST 
2257 SAN JOSE AVE 
65 WAVERLY PL 
339 JUDAH ST 
295 B ORIZABA AVE 
2460 SAN BRUNO AVE 
602 EDDY ST 
634 CLEMENT ST 
1031 IRVING ST 
1315 POWELL ST 
716 KEARNY ST 
3242 SCOTT ST 
1329 COLUMBUS AVE 
655 LARKIN ST 
3608 BALBOA ST 
720 POST ST 
2022 MISSION ST 
5322 GEARY BLVD 
1220 16TH ST 
52.0
55.0
55.0
55.0
56.0
56.0
57.0
57.0
57.0
57.0
57.0
57.0
58.0
59.0
59.0
59.0
59.0
60.0
60.0
60.0


## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!